### PR TITLE
Drop column NumFollowers and NumFollowing

### PR DESCRIPTION
**Pull request by yp05327**
In old design, we had no user visibility, so we stored them as const variables in db.
Since we added user visibility, we should check whether user can see someone, so there was a fix in #20220.
But we didn't notice these unnecessary const variables in db, as they are dynamically calculated by `user_model.GetUserFollowers` and `user_model.GetUserFollowing`.

Discussed in:
https://github.com/go-gitea/gitea/pull/24345#issuecomment-1522835604
https://github.com/go-gitea/gitea/pull/24345#issuecomment-1524872736

### CODE REVIEW

**Review by lunny — DISMISSED** [review]

**Review by delvh — COMMENTED** [review]
What I haven't quite understood:
You only remove in this PR.
Where do you calculate it?
I don't think it's a good idea to extract the calculation into another PR.

**Comment by KN4CK3R** [comment]
The profile page calculates this value always, so nothing is missing. But the API needs that calculation too.

**Comment by yp05327** [comment]
> What I haven't quite understood: You only remove in this PR. Where do you calculate it? I don't think it's a good idea to extract the calculation into another PR.

It is calculated by `user_model.GetUserFollowers` and `user_model.GetUserFollowing`.
In old design, we had no user permission, so we stored them as const variables in db.
Since we added user permission, we should check whether user can see someone, so there was a fix in #20220.
So we have done the calculation, but didn't notice these unnecessary const variables in db, as they are calculated dynamically.

And not only `NumFollowers` and `NumFollowing`, but also `NumWatches`, `NumStars`, `NumForks`... should be calculated dynamically and the stored data in db will be unnecessary I think.

**Review by delvh — APPROVED** [review]

**Comment by yardenshoham** [comment]
Document the deprecation in the PR description

**Comment by yp05327** [comment]
> Document the deprecation in the PR description

Updated.

**Comment by yp05327** [comment]
After finish https://github.com/go-gitea/gitea/pull/24676#issuecomment-1546755533, I will convert this PR into Open state.

File: services/convert/user.go [hunk 1]
```diff
@@ -59,13 +55,31 @@ func toUser(ctx context.Context, user *user_model.User, signed, authed bool) *ap
...
 
+	authed := false
+	signed := false
+	if doer != nil {
+		signed = true
+		authed = doer.ID == user.ID || doer.IsAdmin
+
+		_, followersCount, err := user_model.GetUserFollowers(ctx, user, doer, db.ListOptions{ListAll: true})
+		if err != nil {
+			return nil
```
Comment (yp05327) [inline services/convert/user.go:69]: Not sure whether it is right here.
Same in L76.

Comment (lunny) [reply services/convert/user.go:69]: We should not find all followers here, we just need a number. It will waste unnecessary CPU and Memory.

File: services/convert/user.go [hunk 2]
```diff
@@ -59,13 +54,31 @@ func toUser(ctx context.Context, user *user_model.User, signed, authed bool) *ap
...
+		}
+		result.Followers = int(followersCount)
+		_, followingCount, err := user_model.GetUserFollowing(ctx, user, doer, db.ListOptions{ListAll: true})
```
Comment (lunny) [inline services/convert/user.go]: The same as above.

File: services/convert/user.go (hunk 1 above)
Comment (yp05327) [reply services/convert/user.go:69]: What do you mean `we just need a number`.
Just set a fake number ( 0 ) , or set `Page = 1` and `PageSize = xx` in ListOptions.

File: models/user/user.go [hunk 3]
```diff
@@ -355,6 +353,44 @@ func GetUserFollowing(ctx context.Context, u, viewer *User, listOptions db.ListO
 	return users, count, err
 }
 
+// GetUserFollowersCount returns count of user's followers.
+func GetUserFollowersCount(ctx context.Context, u, viewer *User, listOptions db.ListOptions) (int64, error) {
+	sess := db.GetEngine(ctx).
```
Comment (lunny) [inline models/user/user.go]: Maybe all of this should be extracted as a function, so it could be shared with `GetUserFollowers`.

File: models/user/user.go [hunk 4]
```diff
@@ -355,6 +353,44 @@ func GetUserFollowing(ctx context.Context, u, viewer *User, listOptions db.ListO
...
+	if listOptions.Page != 0 {
+		sess = db.SetSessionPagination(sess, &listOptions)
+		count, err := sess.Count()
+		return count, err
+	}
+
+	count, err := sess.Count()
+	return count, err
+}
+
```
Comment (lunny) [inline models/user/user.go]:
```suggestion
```

File: models/user/user.go [hunk 5]
```diff
@@ -355,6 +353,44 @@ func GetUserFollowing(ctx context.Context, u, viewer *User, listOptions db.ListO
...
+		Where("follow.user_id=?", u.ID).
+		And("`user`.type IN (?, ?)", UserTypeIndividual, UserTypeOrganization).
+		And(isUserVisibleToViewerCond(viewer))
+
+	if listOptions.Page != 0 {
+		sess = db.SetSessionPagination(sess, &listOptions)
+		count, err := sess.Count()
+		return count, err
+	}
+
```
Comment (lunny) [inline models/user/user.go]:
``` suggestion
```

File: models/user/user.go (hunk 4 above)
Comment (yp05327) [reply models/user/user.go:374]: maybe only remove L367 and L368 is enough?

File: models/user/user.go (hunk 3 above)
Comment (yp05327) [reply models/user/user.go]: Done.

File: services/convert/user.go [hunk 6]
```diff
@@ -41,12 +35,12 @@ func ToUserWithAccessMode(ctx context.Context, user *user_model.User, accessMode
...
 	}
-	return toUser(ctx, user, accessMode != perm.AccessModeNone, false)
+	return toUser(ctx, user, nil, accessMode)
 }
 
 // toUser convert user_model.User to api.User
-// signed shall only be set if requester is logged in. authed shall only be set if user is site admin or user himself
-func toUser(ctx context.Context, user *user_model.User, signed, authed bool) *api.User {
+// accessMode is only used if doer is nil
+func toUser(ctx context.Context, user, doer *user_model.User, accessMode perm.AccessMode) *api.User {
```
Comment (lunny) [inline services/convert/user.go:43]: Looks like `accessMode` is always `perm.AccessModeNone`, maybe we should remove it.

Comment (yp05327) [reply services/convert/user.go:43]: If `doer` is nil, we use `accessMode` to check whether provide `Email` info.

**Comment by yp05327** [comment]
image
It seems that there is a bug in sql query.

File: models/user/user.go [hunk 7]
```diff
@@ -311,14 +310,27 @@ func (u *User) GenerateEmailActivateCode(email string) string {
...
 		Select("`user`.*").
 		Join("LEFT", "follow", "`user`.id=follow.user_id").
 		Where("follow.follow_id=?", u.ID).
 		And("`user`.type=?", UserTypeIndividual).
 		And(isUserVisibleToViewerCond(viewer))
+}
+
+func searchUserFollowing(ctx context.Context, u, viewer *User) *xorm.Session {
+	return db.GetEngine(db.DefaultContext).
+		Select("`user`.*").
```
Comment (lunny) [inline models/user/user.go]: Select should only be used with `find` but not `count`.

File: models/user/user.go [hunk 8]
```diff
@@ -311,14 +310,27 @@ func (u *User) GenerateEmailActivateCode(email string) string {
 	return code
 }
 
-// GetUserFollowers returns range of user's followers.
-func GetUserFollowers(ctx context.Context, u, viewer *User, listOptions db.ListOptions) ([]*User, int64, error) {
-	sess := db.GetEngine(ctx).
+func searchUserFollowers(ctx context.Context, u, viewer *User) *xorm.Session {
+	return db.GetEngine(ctx).
 		Select("`user`.*").
```
Comment (lunny) [inline models/user/user.go]: Select should only be used with find but not count.

File: models/user/user.go [hunk 9]
```diff
@@ -312,14 +311,25 @@ func (u *User) GenerateEmailActivateCode(email string) string {
...
+	return db.GetEngine(ctx).
+		Join("LEFT", "follow", "`user`.id=follow.follow_id").
+		Where("follow.user_id=?", u.ID).
+		And("`user`.type IN (?, ?)", UserTypeIndividual, UserTypeOrganization).
+		And(isUserVisibleToViewerCond(viewer))
+}
+
+// GetUserFollowers returns range of user's followers.
+func GetUserFollowers(ctx context.Context, u, viewer *User, listOptions db.ListOptions) ([]*User, int64, error) {
+	sess := searchUserFollowers(ctx, u, viewer)
```
Comment (lunny) [inline models/user/user.go]: Since `Select("`user`.*").` removed from `searchUserFollowers`, we need it here to reduce queries from joined tables.

File: models/migrations/migrations.go [hunk 10]
```diff
@@ -493,8 +493,10 @@ var migrations = []Migration{
 	NewMigration("Add is_internal column to package", v1_20.AddIsInternalColumnToPackage),
 	// v257 -> v258
 	NewMigration("Add Actions Artifact table", v1_20.CreateActionArtifactTable),
-	// v258 -> 259
+	// v258 -> v259
 	NewMigration("Add PinOrder Column", v1_20.AddPinOrderToIssue),
+	// v259 -> v290
```
Comment (lunny) [inline models/migrations/migrations.go]: 259 + 1 = 260 :)

File: services/convert/user.go [hunk 11]
```diff
@@ -59,13 +53,31 @@ func toUser(ctx context.Context, user *user_model.User, signed, authed bool) *ap
...
+			return nil
+		}
+		result.Followers = int(count)
+		count, err = user_model.GetUserFollowingCount(ctx, user, doer)
+		if err != nil {
+			return nil
+		}
+		result.Following = int(count)
+	} else if accessMode != perm.AccessModeNone {
+		signed = true
```
Comment (lunny) [inline services/convert/user.go:78]: An non-login user has read premission to a public repository.

Comment (yp05327) [reply services/convert/user.go:78]: What is the problem? The logic is same to the old code.

File: services/convert/user.go [hunk 12]
```diff
@@ -17,13 +17,7 @@ func ToUser(ctx context.Context, user, doer *user_model.User) *api.User {
...
 		return nil
 	}
-	authed := false
-	signed := false
-	if doer != nil {
-		signed = true
-		authed = doer.ID == user.ID || doer.IsAdmin
-	}
-	return toUser(ctx, user, signed, authed)
+	return toUser(ctx, user, doer, perm.AccessModeNone)
```
Comment (lunny) [inline services/convert/user.go:20]: Looks like it's unrelated changes?
